### PR TITLE
Avoid negative integer overflow when filesize < io_->tell()

### LINKED ISSUE #791: Integer overflow, leading to very large allocation in WebPImage::decodeChunks()
There is a bug at [webpimage.cpp:515](https://github.com/Exiv2/exiv2/blob/55dfdb9ab007d6e6d1d06240048690eaac23b305/src/webpimage.cpp#L515). If `filesize < io_->tell()` then the subtraction overflows and the bounds check on `size` is ineffective.

There is an additional problem in this function, which is that the error conditions are not checked after the calls to `io_->read`. Calls to `io_->read` should always be followed by calls to `io_->error()` and `io_->eof()`, to make sure that the buffer was read successfully. This means that the code does not notice that the file size is much smaller than the value of `size` would suggest. This has enabled me to write a PoC which causes exiv2 to run for almost a minute, scanning 4GB of uninitialized memory.

Here is the PoC, which I have tested on master ([55dfdb9](https://github.com/Exiv2/exiv2/commit/55dfdb9ab007d6e6d1d06240048690eaac23b305)): 
poc1. You can run the PoC like this to see the long-running behavior:

```
$ time exiv2 poc1.jpg  
File name       : poc1.jpg
File size       : 28 Bytes
MIME type       : image/webp
Image size      : 0 x 0
poc1.jpg: No Exif data found in the file

real    0m52.738s
user    0m50.621s
sys     0m2.108s
```

Or if you use `ulimit` to lower the maximum allocation size to less than 4GB then exiv2 will crash:

```
$ ulimit -v 4000000
$ exiv2 poc1.jpg
Uncaught exception: std::bad_alloc
```

### PR DESCRIPTION
**Pull request by kevinbackhouse**
This fixes #791.

I added a new utility function to `BasicIO` to make it more convenient to read `rcount` bytes and throw an exception if not enough bytes are available. I called it `readOrThrow` because I couldn't think of a good name for it.

### CODE REVIEW

**Review by piponazo — COMMENTED** [review]
Very nice idea! Ideally we could start replacing step by step the old usages of `read` by this new `readOrThrow` so that we are always sure about checking the whether the **read** operation succeed.

I would even vote for not having the 3rd argument and always throw the same kind of error (`Exiv2::kerCorruptedMetadata`) . 

I would also like to see @D4N 's opinion in this matter.

**Comment by D4N** [comment]
I like this new API and will follow up with a proper review hopefully in the next few days. I have two comments:
- imho this is quite an API change for the 0.27 maintenance branch, not sure if we should introduce this
- `readOrThrow` should imho replace `read` completely, except for rare cases where we don't want an exception.

**Comment by clanmills** [comment]
I agree with @D4N Please don't make API changes on 0.27-maintenance.

In fact, I should open an issue on myself to investigate if we made API changes in 0.27.1 and I'll update the release process to verify that we haven't introduced changes.  The aim of the v0.27-dots is to avoid API changes so security fixes can be provided by swapping DLLs without recompiling applications.

**Comment by piponazo** [comment]
I also agree with @D4N comments:

1. I did not think about the API change, but it is true. We should not change the API in 0.27-maintenance.
2. We could keep the `read` function with the same signature but changing the implementation for always checking if the read was successful or not. In principle it should not break anything and it should make the code safer against some of these cooked POCs.

**Comment by clanmills** [comment]
Can we use a macro or template to implement `readOrThrow()`, then we don't change the API?

**Comment by kevinbackhouse** [comment]
I changed this to avoid the API change by making `readOrThrow` a static function in `webpimage.cpp`. I also posted #813, which adds the API change on master.

**Comment by piponazo** [comment]
Hi @kevinbackhouse , as I commented in #813 the build on windows is not passing:

```
  C:\projects\exiv2-wutfp\src\webpimage.cpp(524): warning C4018: '<=': signed/unsigned mismatch [C:\projects\exiv2-wutfp\build\src\exiv2lib.vcxproj]
```
If you do not have a windows development environment let me know and I'll take a look at it.

**Comment by kevinbackhouse** [comment]
@piponazo: I have pushed a commit to fix this. The problem is that `long` is only 32 bits on Windows. Which is a problem because it means that a `uint32_t` could overflow when cast to `long`. So I have changed all the types to `long` but also added extra checks for overflow.

File: src/webpimage.cpp [hunk 1]
```diff
@@ -489,15 +498,16 @@ namespace Exiv2 {
         DataBuf chunkId(5);
         chunkId.pData_[4] = '\0' ;
 
-        io_->read(data, WEBP_TAG_SIZE * 3);
+        readOrThrow(*io_, data, WEBP_TAG_SIZE * 3, Exiv2::kerCorruptedMetadata);
 
-        const uint32_t filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
-        enforce(filesize <= io_->size(), Exiv2::kerCorruptedMetadata);
+        const long filesize = Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 8;
```
Comment (D4N) [inline src/webpimage.cpp]: See my comment in #813, use `Safe::add`.

File: src/webpimage.cpp [hunk 2]
```diff
@@ -508,10 +518,12 @@ namespace Exiv2 {
...
         chunkId.pData_[4] = '\0' ;
-        while ( !io_->eof() && (uint64_t) io_->tell() < filesize) {
-            io_->read(chunkId.pData_, WEBP_TAG_SIZE);
-            io_->read(size_buff, WEBP_TAG_SIZE);
-            const uint32_t size = Exiv2::getULong(size_buff, littleEndian);
+        while (!io_->eof() && io_->tell() < filesize) {
+            readOrThrow(*io_, chunkId.pData_, WEBP_TAG_SIZE, Exiv2::kerCorruptedMetadata);
+            readOrThrow(*io_, size_buff, WEBP_TAG_SIZE, Exiv2::kerCorruptedMetadata);
+            const long size = Exiv2::getULong(size_buff, littleEndian);
+            enforce(0 <= size, Exiv2::kerCorruptedMetadata);
```
Comment (D4N) [inline src/webpimage.cpp]: See my comment in #813, this is relying on undefined behavior. To properly check this, you must do something like this:
``` C++
const uint32_t size = Exiv2::getULong(size_buff, littleEndian);
enforce(size < std::numeric_limits<long>::max(), Exiv2::kerCorruptedMetadata);
const long long_size = size;
```

**Review by D4N — DISMISSED** [review]
Overall this looks good to me for `0.27-maintenance`, I've added a few notes concerning the overflow checks.

File: src/webpimage.cpp (hunk 1 above)
Comment (kevinbackhouse) [reply src/webpimage.cpp]: done

File: src/webpimage.cpp (hunk 2 above)
Comment (kevinbackhouse) [reply src/webpimage.cpp]: done

**Review by piponazo — DISMISSED** [review]

**Review by piponazo — APPROVED** [review]

**Pull request merged**